src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -584,10 +586,19 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
                 let msg = if !has_fork && partial { "partially " }
                           else if has_fork && !has_common { "collaterally "}
                           else { "" };
-                let err = struct_span_err!(
+                let mut err = struct_span_err!(
                     self.tcx.sess, use_span, E0382,
                     "{} of {}moved value: `{}`",
                     verb, msg, nl);
+                match lp.ty.sty {
+                    ty::TypeVariants::TyClosure(id, _) => {
+                        if let Ok(ty::ClosureKind::FnOnce) =
+                           ty::queries::closure_kind::try_get(self.tcx, DUMMY_SP, id) {
+                            err.help("closure was moved because it only implements `FnOnce`");
```

Comment:
Hmm. I'm torn. This message is definitely an improvement. I wonder if we can't do better. One thing that would be **really** nice is if we could show **why** the closure implements `FnOnce` (i.e., because it has a move from the environment, typically).

We would then say something like:

- closure cannot be invoked twice because it moves the variable `dict` out of its environment

Computing this would require modifying `upvar.rs` to make the information available. But it seems worthwhile (it's been on my to-do list for a while). Possibly should be deferred to a separate PR, I'm not sure.